6 list persistance

tycebrown opened the pull request:
Here is a design overview:

- Instead of directly storing a List<Item>, I've abstracted that into a class called ListModel, which can be directly put into an isar database.
- Isar database work is abstracted into a static class called ListsDatabaseManager. 

Some design questions/concerns:

- At the current moment, the list model (and its items) are loaded "eagerly", so all the lists and their items are loaded at the start. Should the items and list model's metadata (such as the name of the list model) be loaded separately? Do you think an "eager" loading would seriously slow down the app? 
- All updates are eager (I couldn't find any way to schedule updates to happen at the termination of the app), so a slight change leads to the whole list model being re-put into the database. Would this slow down the app _noticeably_?
- Are there any sloppy names that need to be fixed (like ListsDatabaseManager)?

tycebrown committed:
Raw, unclean code for list persistence(not to be reviewed)

tycebrown committed:
Cleaned up code, edited ListWidget doc comment

tycebrown committed:
Small Item class doc comment change

tycebrown committed:
Renamed listModel field "name" to "title"

Pertempto commented:
> Here is a design overview:
> 
> * Instead of directly storing a List, I've abstracted that into a class called ListModel, which can be directly put into an isar database.
> * Isar database work is abstracted into a static class called ListsDatabaseManager.
> 
> Some design questions/concerns:
> 
> * At the current moment, the list model (and its items) are loaded "eagerly", so all the lists and their items are loaded at the start. Should the items and list model's metadata (such as the name of the list model) be loaded separately? Do you think an "eager" loading would seriously slow down the app?
> * All updates are eager (I couldn't find any way to schedule updates to happen at the termination of the app), so a slight change leads to the whole list model being re-put into the database. Would this slow down the app _noticeably_?
> * Are there any sloppy names that need to be fixed (like ListsDatabaseManager)?

I think the current "eager" system should be fine for now. We can optimize later if we need. For `ListsDatabaseManager`, why not call it just `DatabaseManager` or `DbManager`?

tycebrown committed:
Renamed ListsDatabaseManager -> DatabaseManager

Pertempto commented on lib/model/lists_database_manager.dart:
```diff
@@ -0,0 +1,28 @@
+  static Future<List<ListModel>> getAllListModels() async =>
+      await isar.listModels.where().findAll();
+
+  static Future<void> updateListModel(ListModel which) async =>
```
My question is: why do we have to write all the list models to isar on every update? There is no way to update one specific item?

tycebrown replied:
What I had in mind was that the main function would call `getAllListModels()` and pass the returned list of `ListModel`'s to the app widget. Then, after the user selects a list, the app widget would pass the selected list to a `ListWidget`. 

So, `getAllListModels()` would be called only once at the beginning of the program (or somewhere near the beginning).

In the ListWidget, whenever the user changes the list, the updated list is written immediately to the database, but only the changed list is `put()` to the `isar.listModels` database. (In isar, each time you need to write to a database, the transactions have to be placed inside an `isar.writeTxn(() async { /* the database writes */ })` call).

So, with the current DatabaseManager, you can't _get_ a single list item, but you can _update_ a single item.

Did I answer your question?

Pertempto replied:
Kind of. From how your code reads, it looks like you are updating an _entire_ list model, not just one item. Maybe I am misunderstanding?

tycebrown replied:
Ah. You are understanding. I see what you mean. My code is updating the entire list model. 

It would be possible to update only a specific item if I tweaked the code a bit. Will do.

tycebrown replied:
Sorry, in my first answer, I mispoke. I meant to say you can't get a single list _model_, but you can update a single model.

Pertempto replied:
Ok, that makes much more sense!

tycebrown committed:
Renamed lists_database_manager (dart file) to database_manager

Pertempto commented on lib/model/lists_database_manager.dart:
```diff
@@ -0,0 +1,28 @@
+import 'package:flutter/foundation.dart';
```
Don't forget to rename this file to `database_manager.dart`.

tycebrown committed:
Made it possible to add/remove/update a single item.

tycebrown committed:
Revert "Made it possible to add/remove/update a single item."

This reverts commit bd8346b9a513cc96b68ca66e3f263406e1c1a363.

tycebrown committed:
Now, single items can be updated and written into the database. Code needs a little cleaning.

tycebrown committed:
A little code sweeping.

Removed createListModel and createItem and replaced by with putListModel and putItem.

add was renamed to addOrUpdate

Finally, I change the remove and addOrUpdate functions in such a way so that they are now responsible for deleting and putting items to the database as well as creating links to them.

Love to hear thoughts.

tycebrown committed:
Sorry, should be ready now for review.

Pertempto commented on lib/model/list_model.dart:
```diff
@@ -0,0 +1,41 @@
+
+  void init() => items.loadSync();
+
+  Future<void> addOrUpdate(Item newItem) async {
```
Can you split this into two separate `add` and `update` methods? I think it would make more sense.

tycebrown replied:
That makes sense. Do you think the `update` method should check to see if the `Item` to be updated is in `this.items` before it updates it? Or should it be assumed (a precondition)?

Pertempto replied:
That could be good. If the check failed, you should throw some kind of error so that the problem shows up to developers.

Pertempto commented on lib/model/list_model.dart:
```diff
@@ -0,0 +1,41 @@
+  }
+
+  /// For use by the DatabaseManager only. Use `DatabaseManager.updateListModelItems(listModel)` instead
+  Future<void> updateItems() async => await items.save();
```
I don't like this strange circular dependency between `DatabaseManager` and `ListModel`. I think you can remove this method.

Pertempto commented on lib/model/database_manager.dart:
```diff
@@ -0,0 +1,39 @@
+        ..forEach((listModel) => listModel.init());
+
+  static Future<void> updateListModelItems(ListModel which) async =>
+      await isar.writeTxn(() async => await which.updateItems());
```
```suggestion
      await isar.writeTxn(() async => await which.items.save());
```

Pertempto commented on lib/model/database_manager.dart:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  static Future<ListModel> putListModel(ListModel listModel) async {
+    // final newListModel = ListModel.fromTitle(name);
```
What is this line for?

tycebrown replied:
Oh, that comment is residue from old code. Will clean.

Pertempto reviewed (changes requested):
Functionality is still working great! Just a few code-cleanup things I saw.

tycebrown committed:
Removed my weird circular dependency and piece of commented out code

tycebrown committed:
Ready for review: added error checking for the update function

Pertempto reviewed (approved):
The code looks good!
